### DISCUSSION TITLE
How to prepare for worst case scenario if website goes down?

### DISCUSSION THREAD

**Discussion by dmusican** [pybricks/support#1001] [Q&A] [answered]
Hi folks,

I've started playing around with Pybricks on Spike Prime, and it's fantastic. I run a 3-week summer program, and I would very much like to switch to Pybricks from the standard LEGO platform.

Our program consists of students who come to stay at our campus from around the world, and so having our platform become unusable mid-program would be a major disaster for us. There's no reason to believe that would happen, but I'm trying to think through that worst-case scenario, just in case. I have had online course materials (e.g. textbooks) get pulled from the web in the middle of a course I've offered, so the concern isn't entirely academic. (Ok, bad pun, academic is what I do.)

One question / concern I have is with regards to what happens if the website goes down or if the entire project disappears. With the standard LEGO software, I've got it installed locally, so even if the LEGO corporation and its entire online presence disappeared overnight, it would have no effect on the locally installed software that I have, and I keep a local copy of the install executables if I need them.

In the case of Pybricks, it lives entirely in the browser, and so I have less of a sense of how robust it is. I'm aware that you can install Pybricks as an app and it works offline. I've tried that, and as far as I can tell in my limited testing, it works great! But I don't understand the technical details regarding how offline web apps works or how robust they are. So, some questions:

- In offline mode, can I install firmware if need be? Can I reinstall the original LEGO firmware if I need to?
- If the website went down and I needed to get Pybricks to work on a new computer that hasn't installed it as an app locally, what options do I have? Is there any way I can locally store files (or even host a local web server?) that would enable new local installs?
- What else am I not thinking about, regarding how to handle if online connections to code.pybricks.com stopped working for some reason?

Thanks for bearing with me on these questions. The Pybricks developers are amazingly responsive and invested in the project, so the likelihood of this happening all seems vanishing small. But I hope the fact that I'm a little skittish about a web-based app for a high-stakes situation makes sense, and so I appreciate any thoughts that you have. Thank you!

**Comment by dlech** [comment, marked as answer]
> In offline mode, can I install firmware if need be? Can I reinstall the original LEGO firmware if I need to?

Yes, everything including flashing the Pybricks firmware and the official LEGO firmware restore works in offline mode.

> If the website went down and I needed to get Pybricks to work on a new computer that hasn't installed it as an app locally, what options do I have?

Pybricks Code is fully open source, so you can build and host it yourself. You can learn more about it [here](https://github.com/pybricks/pybricks-code/blob/master/CONTRIBUTING.md). For hosting, the only requirement is that it uses https, not http. 

> What else am I not thinking about, regarding how to handle if online connections to code.pybricks.com stopped working for some reason?

Pybricks Code is a Progressive web app and works offline through a special browser caching feature that is part of Web Workers. So once you use it once, it is on your computer until the browser cache is cleared. So unless you have a really full hard drive that triggers an automatic browser cache clear or someone decides to delete the entire browser data, you are not likely to lose it.

**Reply by dmusican** [reply]
Many thanks, this is all fantastically helpful. I appreciate the amazingly fast timing on the response as well.
